File: Task3/01_chunking.js
const fs = require("fs");
const path = require("path");
const { RecursiveCharacterTextSplitter } = require("@langchain/textsplitters");

const KB_PATH = path.join(__dirname, "..", "Task2", "knowledge_base");
const OUT_FILE = path.join(__dirname, "output", "chunks.jsonl");

// ============================================================
// Параметры чанкинга:
// - chunkSize  = 800 символов ≈ 500–1000 токенов (100–300 слов)
// - chunkOverlap = 150 символов — перекрытие для сохранения контекста
// - Минимальный размер чанка: 30 символов (фильтруем заголовки-одиночки)
// ============================================================
const CHUNK_SIZE = 800;
const CHUNK_OVERLAP = 150;
const MIN_CHUNK_LENGTH = 30;

async function main() {
  if (!fs.existsSync(KB_PATH)) {
    console.error(`❌  Папка ${KB_PATH} не найдена. Сначала запустите Task2/02_replace.js`);
    process.exit(1);
  }

  const splitter = new RecursiveCharacterTextSplitter({
    chunkSize: CHUNK_SIZE,
    chunkOverlap: CHUNK_OVERLAP,
  });

  const files = fs.readdirSync(KB_PATH).filter((f) => f.endsWith(".txt")).sort();
  console.log(`📂  Найдено ${files.length} файлов в ${KB_PATH}\n`);

  // Создаём выходную папку
  const outDir = path.dirname(OUT_FILE);
  if (!fs.existsSync(outDir)) {
    fs.mkdirSync(outDir, { recursive: true });
  }

  let chunkIdGlobal = 0;
  const outStream = fs.createWriteStream(OUT_FILE, { encoding: "utf-8" });

  for (const file of files) {
    const filePath = path.join(KB_PATH, file);
    const text = fs.readFileSync(filePath, "utf-8");

    const chunks = await splitter.splitText(text);

    let localId = 0;
    for (const chunk of chunks) {
      // Пропускаем слишком короткие чанки (заголовки-одиночки)
      if (chunk.trim().length < MIN_CHUNK_LENGTH) {
        continue;
      }

      // Находим позицию чанка в оригинальном тексте (char_offset)
      const charOffset = text.indexOf(chunk.slice(0, 50));

      const record = {
        id: chunkIdGlobal,
        text: chunk,
        metadata: {
          source: file,
          chunk_id: localId,
          char_offset: charOffset >= 0 ? charOffset : null,
          char_length: chunk.length,
          word_count: chunk.split(/\s+/).length,
        },
      };
      outStream.write(JSON.stringify(record, null, 0) + "\n");
      chunkIdGlobal++;
      localId++;
    }

    console.log(`✅  ${file}: ${localId} чанков`);
  }

  outStream.end(() => {
    console.log(`\n🏁  Чанкинг завершён. Создано чанков: ${chunkIdGlobal}`);
    console.log(`📄  Файл сохранён: ${OUT_FILE}\n`);
  });
}

main();

File: Task3/01_chunking.py
import json
import os
import sys

from langchain_text_splitters import RecursiveCharacterTextSplitter

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
KB_PATH = os.path.join(SCRIPT_DIR, "..", "Task2", "knowledge_base")
OUT_FILE = os.path.join(SCRIPT_DIR, "output", "chunks.jsonl")

# ============================================================
# Параметры чанкинга:
# - chunk_size  = 800 символов ≈ 500–1000 токенов (100–300 слов)
# - chunk_overlap = 150 символов — перекрытие для сохранения контекста
# - Минимальный размер чанка: 30 символов (фильтруем заголовки-одиночки)
# ============================================================
CHUNK_SIZE = 800
CHUNK_OVERLAP = 150
MIN_CHUNK_LENGTH = 30


def main() -> None:
    if not os.path.isdir(KB_PATH):
        print(f"❌  Папка {KB_PATH} не найдена. Сначала запустите Task2/02_replace.py")
        sys.exit(1)

    splitter = RecursiveCharacterTextSplitter(
        chunk_size=CHUNK_SIZE,
        chunk_overlap=CHUNK_OVERLAP,
    )

    files = sorted(f for f in os.listdir(KB_PATH) if f.endswith(".txt"))
    print(f"📂  Найдено {len(files)} файлов в {KB_PATH}\n")

    # Создаём выходную папку
    out_dir = os.path.dirname(OUT_FILE)
    os.makedirs(out_dir, exist_ok=True)

    chunk_id_global = 0

    with open(OUT_FILE, "w", encoding="utf-8") as out_stream:
        for file in files:
            file_path = os.path.join(KB_PATH, file)
            with open(file_path, "r", encoding="utf-8") as f:
                text = f.read()

            chunks = splitter.split_text(text)

            local_id = 0
            for chunk in chunks:
                # Пропускаем слишком короткие чанки (заголовки-одиночки)
                if len(chunk.strip()) < MIN_CHUNK_LENGTH:
                    continue

                # Находим позицию чанка в оригинальном тексте (char_offset)
                char_offset = text.find(chunk[:50])

                record = {
                    "id": chunk_id_global,
                    "text": chunk,
                    "metadata": {
                        "source": file,
                        "chunk_id": local_id,
                        "char_offset": char_offset if char_offset >= 0 else None,
                        "char_length": len(chunk),
                        "word_count": len(chunk.split()),
                    },
                }
                out_stream.write(json.dumps(record, ensure_ascii=False) + "\n")
                chunk_id_global += 1
                local_id += 1

            print(f"✅  {file}: {local_id} чанков")

    print(f"\n🏁  Чанкинг завершён. Создано чанков: {chunk_id_global}")
    print(f"📄  Файл сохранён: {OUT_FILE}\n")


if __name__ == "__main__":
    main()
